File: python/taflow/metrics/win_rate.py
"""Profitable-observation frequency metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import WinRate as _Native
from ._input import as_metric_series


class WinRate:
    """Compute strictly positive observations divided by non-zero observations.

    A win is strictly greater than zero, a loss is strictly less than zero,
    and exact zero is a breakeven excluded from the ratio denominator. The
    external oracle is QuantStats 0.0.81 ``win_rate`` with no aggregation and
    ``prepare_returns=False``. TAFlow normalizes QuantStats' arbitrary zero for
    a sample with no non-zero observations to ``None`` because the ratio is
    mathematically undefined. Warm-up ends with the first win or loss; zero is
    a valid result after a loss-only sample. ``len(metric)`` counts every usable
    observation, including breakevens, while omitted NaNs do not advance it.

    ``from_returns`` consumes chronological decimal simple period returns.
    ``from_pnl`` consumes raw, non-cumulative period P&L and deliberately does
    not accept initial equity. ``from_trades`` consumes realized P&L for each
    closed trade. These domains preserve sign classification without conversion
    and are not annualized. Equity and log-return factories are intentionally
    absent. NaNs are omitted by default or rejected with
    ``nan_policy="raise"``; infinities are always rejected, and return-domain
    values below -100% are invalid. ``append``, ``extend``, and ``reset`` are
    fluent. Bulk work releases the GIL; all counting and arithmetic live in Rust.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use a semantic ``from_*`` factory."""
        raise TypeError("use WinRate.from_returns/from_pnl/from_trades")

    @classmethod
    def _create(
        cls,
        values: Any,
        input_mode: str,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "WinRate":
        state = cls.__new__(cls)
        state._state = _Native(input_mode, nan_policy=nan_policy)
        return state.extend(values, column=column)

    @classmethod
    def from_returns(
        cls,
        returns: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "WinRate":
        """Construct from chronological decimal simple period returns."""
        return cls._create(
            returns, "returns", nan_policy=nan_policy, column=column
        )

    @classmethod
    def from_pnl(
        cls,
        pnl: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "WinRate":
        """Construct from raw non-cumulative period P&L without conversion."""
        return cls._create(pnl, "pnl", nan_policy=nan_policy, column=column)

    @classmethod
    def from_trades(
        cls,
        trades: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "WinRate":
        """Construct from realized P&L observations for closed trades."""
        return cls._create(trades, "trades", nan_policy=nan_policy, column=column)

    def append(self, value: float) -> "WinRate":
        """Append one value in the factory-selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "WinRate":
        """Append a series in the selected domain and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return the current win rate, or ``None`` without a win or loss."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "WinRate":
        """Clear observations, preserve input configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return usable observations including zero-valued breakevens."""
        return len(self._state)


__all__ = ["WinRate"]

File: crates/taflow-metrics/src/metrics/win_rate.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Share of decisive observations that are strictly positive.
#[derive(Debug, Clone)]
pub struct WinRate {
    input: MetricInputState,
    wins: usize,
    decisive_observations: usize,
}

impl WinRate {
    /// Construct an empty state for returns, raw period P&L, or closed trades.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if !matches!(
            input_kind,
            MetricInputKind::Returns | MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "win rate requires returns, raw period P&L, or closed-trade P&L",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            wins: 0,
            decisive_observations: 0,
        })
    }

    /// Append one chronological observation and return the rate to date.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            if observation > 0.0 {
                self.wins += 1;
                self.decisive_observations += 1;
            } else if observation < 0.0 {
                self.decisive_observations += 1;
            }
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.input.extend(values, |observation| {
            if observation > 0.0 {
                self.wins += 1;
                self.decisive_observations += 1;
            } else if observation < 0.0 {
                self.decisive_observations += 1;
            }
            Ok(())
        })?;
        Ok(self.value())
    }

    /// Return wins divided by non-zero observations, or `None` without one.
    pub fn value(&self) -> Option<f64> {
        (self.decisive_observations != 0)
            .then(|| self.wins as f64 / self.decisive_observations as f64)
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.wins = 0;
        self.decisive_observations = 0;
    }

    /// Return the number of usable observations, including breakevens.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable observations have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
